### DISCUSSION TITLE
nameof should support Byte[]

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#5006] [General]
``` C#
public static SqlDbType ToSqlDbType(object obj)
{
    if (obj is null)
        throw new ArgumentNullException(nameof(obj));

    var result = SqlDbType.NChar;
    var type = obj.GetType();

    result = type.Name switch
    {
        nameof(Boolean) => SqlDbType.Bit,
        nameof(Byte) => SqlDbType.TinyInt,
        nameof(Int16) => SqlDbType.SmallInt,
        nameof(Int32) => SqlDbType.SmallInt,
        nameof(Single) => SqlDbType.Real,
        nameof(Double) => SqlDbType.Float,
        nameof(String) => SqlDbType.NChar,
        nameof(Guid) => SqlDbType.UniqueIdentifier,
        nameof(XmlReader) => SqlDbType.Xml,
        nameof(Decimal) => SqlDbType.Money,
        nameof(DateTime) => SqlDbType.DateTime2,
        "Byte[]" => SqlDbType.Binary,
        nameof(Object) => SqlDbType.Variant,
        _ => result
    };

    return result;
}
```

**Comment by HaloFour** [comment]
It'd be even better if you could `switch` directly on `Type`:

```cs
public static SqlDbType ToSqlDbType(object obj)
{
    if (obj is null)
        throw new ArgumentNullException(nameof(obj));

    var result = SqlDbType.NChar;
    var type = obj.GetType();

    result = type switch
    {
        typeof(bool) => SqlDbType.Bit,
        typeof(byte) => SqlDbType.TinyInt,
        typeof(short) => SqlDbType.SmallInt,
        typeof(int) => SqlDbType.SmallInt,
        typeof(float) => SqlDbType.Real,
        typeof(double) => SqlDbType.Float,
        typeof(string) => SqlDbType.NChar,
        typeof(Guid) => SqlDbType.UniqueIdentifier,
        typeof(XmlReader) => SqlDbType.Xml,
        typeof(Decimal) => SqlDbType.Money,
        typeof(DateTime) => SqlDbType.DateTime2,
        typeof(byte[]) => SqlDbType.Binary,
        typeof(object) => SqlDbType.Variant,
        _ => result
    };

    return result;
}
```

**Reply by ghost** [reply]
Yeah. But, I cannot use the code to do.
Please see: https://github.com/dotnet/csharplang/issues/5007

**Reply by svick** [reply]
> It'd be even better if you could `switch` directly on `Type`

That's https://github.com/dotnet/csharplang/discussions/356.

**Comment by bernd5** [comment]
Why not simply:
```cs
using System;
using System.Data;
using System.Xml;

public class App{

    public static SqlDbType ToSqlDbType(object obj)
    {
        if (obj is null)
            throw new ArgumentNullException(nameof(obj));

        var result = SqlDbType.NChar;
        result = obj switch
        {
            bool => SqlDbType.Bit,
            byte => SqlDbType.TinyInt,
            short => SqlDbType.SmallInt,
            int => SqlDbType.SmallInt,
            float => SqlDbType.Real,
            double => SqlDbType.Float,
            string => SqlDbType.NChar,
            Guid => SqlDbType.UniqueIdentifier,
            XmlReader => SqlDbType.Xml,
            Decimal => SqlDbType.Money,
            DateTime => SqlDbType.DateTime2,
            byte[] => SqlDbType.Binary,
            object => SqlDbType.Variant,
            _ => result
        };

        return result;
    }
    
    static void Main()
    {
        Console.WriteLine(ToSqlDbType(true));
    }
}
```
(BTW.: the fallback `NChar` is never true)

**Reply by Joe4evr** [reply]
> It's a project on .NET Framwork and C# 8.0

https://github.com/dotnet/csharplang/issues/5007#issuecomment-894019775

**Reply by bernd5** [reply]
Why not updating to a newer version of csharp? CSharp 8.0 is fixed - every new feature would require a compiler / compiler-setting update.

For this is no special runtime support needed. Just look at the lowered code: [code](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBAGABAlAOgBEIAXCAbkx3yIA0wAbajGgZnyVwEEAHPgG9MmXGPycCANlwBlAI5NiwACoBPPgFNcKgPYKlqjZoAUu4ACtNAYzK5zFgJSjxwjOI+4oAM1xnLXgDOuDAArkxMzu6eMWQAFgBOugDuIZqpPAkA5qFgmjBkAHLhTACiAB7WmnxkULowJjAQebre/k6OrC4xAG4QCbgJmoHhdgC8corK6lqEhQDCcf2sMWJDI0zj9gGByVBk1nHdnm6rnsC6uky4YwB8k4YzmoQAQvsANMdnwGpk2ncPabGQgqWBqACSBU+0TOYkCcV0CXG9wMQNmskgEUhZGhsPEsGRgKM6MxTGxuLxuG8TF05BuKKmxOeACVNBAmBS8QATXShYBMf4Mx7AgBiNPInNhBDwANRTLmi36krOAHFQlAufSiU9CABVGBQeShTTgrn5WreKCaBLK1aMJisiBmgayxk6+22mLEGxQTFauU6gCy9U0ak9nlIf1BeX9buBkc00c0SHDHh+fwA2gBdWPC2ZvJoJMNfVYOGyEgPAgBq/SgEChJZiAH0tetRo3cABfLowzwIADsg2GoxW4k7x2O0nwABZcIHoA0oh5TqsCABOEx6StaExkBLGxydY7jjCdoA===)

**Comment by ZacharyPatten** [comment]
If https://github.com/dotnet/csharplang/discussions/2505#discussioncomment-1138842 was a thing then maybe we could use a source generator to gen a `static` `const` `string` for each desired type. Instead of `nameof(byte[])`  could be `(byte[]).nameof`.

Not saying I prefer that... but just a thought.
